src/cantools/subparsers/monitor.py

```diff
@@ -338,74 +350,68 @@ def process_user_input_filter(self, key):
                 self._filter_cursor_pos += 1
 
         self.compile_filter()
-        self._filtered_sorted_message_names = []
 
-        for name in self._formatted_messages:
-            self.insort_filtered(name)
+        # reformat all messages
+        self._filtered_sorted_message_names.clear()
+        self._message_signals.clear()
+        self._formatted_messages.clear()
+        for msg in self._raw_messages.values():
+            self.try_update_message(msg)
 
         self._modified = True
 
-    def try_update_message(self):
-        message = self._queue.get_nowait()
-        frame_id = message.arbitration_id
-        data = message.data
-        timestamp = message.timestamp
-
+    def try_update_message(self, raw_message: can.Message) -> MessageFormattingResult:
         if self._basetime is None:
-            self._basetime = timestamp
+            self._basetime = raw_message.timestamp
 
-        timestamp -= self._basetime
-        self._received += 1
+        data = raw_message.data
+        timestamp = raw_message.timestamp - self._basetime
 
         try:
-            message = self._dbase.get_message_by_frame_id(frame_id)
+            message = self._dbase.get_message_by_frame_id(raw_message.arbitration_id) # type: ignore[union-attr]
         except KeyError:
-            self._discarded += 1
-            return
+            return MessageFormattingResult.UnknownMessage
 
         name = message.name
         try:
             if message.is_container:
                 self._try_update_container(message, timestamp, data)
-                return
+            else:
+                if len(data) < message.length:
+                    self._update_message_error(timestamp, name, data, f'{message.length - len(data)} bytes too short')
+                    return MessageFormattingResult.DecodeError
 
+                if message.is_multiplexed():
+                    name = format_multiplexed_name(message,
```

Comment:
if you change `format_message()` to take the decoded data (good idea, btw), you should probably also adapt `format_multiplexed_name()` accordingly. (maybe this function should be renamed as well. `format_multiplexed_message()`?)

Reply:
I have adapted `format_multiplexed_name` and kept the function name, as the function is returning `name` - is it okay?

Just found out that decoding containers was not showing correct data even on master - already decoded `cdata` should be used instead of decoding that was hidden in `format_message`. Fixed newly introduced test and added one more test to check encoding of multiplexed messages in a container.

Currently we are having little-bit duplicated code for formatting normal messages and containerized ones. The container version is also not respecting single-line configuration. I could refactor it and union behavior in following PR if the single-line should be respected.

Reply:
> Currently we are having little-bit duplicated code for formatting normal messages and containerized ones. The container version is also not respecting single-line configuration. I could refactor it and union behavior in following PR if the single-line should be respected.

that would be great. (when I implemented support for containers I simply did not test the single line stuff because without signal filtering this is basically useless for me...)